### PR TITLE
 Gracefully drop mismatching packets during batch aggregation

### LINKED ISSUE #118: Gracefully drop mismatching packets during batch aggregation
Some email discussion highlighted that currently, if we have a mismatch between the packet uuids of the batch and both validation shares, the entire batch is dropped. We should instead drop only the packets that are missing from one of the sources, and add the mismatching uuids to the output_uuid file.

### PR DESCRIPTION
**Pull request by tgeoghegan**
This PR resolves #118. The first commit enhances the sample generation so it can drop every nth packet from the PHA batch and every mth from the facilitator batch, as well as recording expected sums and dropped packet UUIDs. The second commit reimplements aggregation such that we no longer discard whole batches if any individual validation packet is missing.

### CODE REVIEW

**Comment by tgeoghegan** [comment]
Besides the tests in this PR, I was able to validate that data share processors with these changes can ingest and aggregate batches from our partners. I'm keen to release a facilitator container image containing this change to the NCI, so jsha and I agreed to merge this change to unblock the release and proceed with his review feedback asynchronously.

**Review by jsha — COMMENTED** [review]
As a general piece of feedback, I'd recommend more use of type annotations (makes reviewing much easier, particularly in GitHub where I can't use rust-analyzer to show me the types of things). I'd also recommend trying to un-nest things much more.

Anyhow, all of the below is style nits - AFAICT the functionality looks correct.

File: facilitator/src/aggregation.rs [hunk 1]
```diff
@@ -204,71 +210,65 @@ impl<'a> BatchAggregator<'a> {
...
-                &ingestion_packet,
+            // Ignore duplicate packets
+            if processed_ingestion_packets.contains(&ingestion_packet.uuid) {
+                info!("ignoring duplicate packet {}", ingestion_packet.uuid);
+                continue;
+            }
+
+            // Make sure we have own validation and peer validation packets
+            // matching the ingestion packet.
+            let peer_validation_packet = match get_validation_packet(
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply facilitator/src/aggregation.rs:251]: This isn't the first time you've pointed out that multiline values in `match` statements are confusing, and you're right. I'll break these into two variable assignments for clarity.

File: facilitator/tests/integration_tests.rs [hunk 2]
```diff
@@ -2,26 +2,39 @@ use chrono::NaiveDateTime;
...
+}
+
+#[test]
+fn inconsistent_ingestion_batches() {
+    // Have sample generation drop every third and every fourth packet from the
+    // PHA and facilitator ingestion batches, respectively.
+    end_to_end_test(Some(3), Some(4))
+}
+
+fn end_to_end_test(drop_nth_pha: Option<usize>, drop_nth_facilitator: Option<usize>) {
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply facilitator/tests/integration_tests.rs:36]: The two functions above it are `#[cfg(test)]`. This function implements the common test logic for both.

File: facilitator/src/idl.rs [hunk 3]
```diff
@@ -743,6 +756,18 @@ impl Packet for ValidationPacket {
     }
 }
 
+impl Ord for ValidationPacket {
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply facilitator/src/idl.rs:759]: Good catch -- in fact I'd already done something like this for the `Packet::schema()` method

File: facilitator/tests/integration_tests.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsha) [reply facilitator/tests/integration_tests.rs:36]: The two functions above `#[test]`, which says "this is a test." This should be `#[cfg(test)]`, which says "this should only be compiled when running tests.

File: facilitator/src/idl.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply facilitator/src/idl.rs:759]: Looking at this further, the `impl Ord` and `impl PartialOrd`s were a leftover from an iteration where I was sorting all the packet lists. That's no longer the case, so we don't need those implementations on either the trait or the concrete type, so I removed them altogether.

**Pull request merged**